File: MIDTERM-LAB-3/MP02/MP02.java
// ============================================================
// MP02 - Display First 10 Rows of Dataset
// Programmer: EYAS, SOPHIA WELYNNE S.
// Course: Programming 2 | BSIT GD - 1st Year
// Description: This program reads a CSV file and displays
//              the first 10 data rows in a readable format,
//              along with the column headers of the dataset.
// ============================================================

import java.io.*;        // For reading files
import java.util.*;      // For ArrayList and Scanner

public class MP02 {

    public static void main(String[] args) {

        // Scanner is used to get keyboard input from the user
        Scanner scanner = new Scanner(System.in);

        // Ask the user to type the path to the CSV file
        System.out.print("Enter the CSV dataset file path: ");
        String filePath = scanner.nextLine(); // Store what the user typed

        // This list will hold the first 10 data rows we find
        ArrayList<String[]> first10Rows = new ArrayList<>();

        // 'headerColumns' will store the column names from the header row
        String[] headerColumns = null;

        // 'dataRowCount' counts how many actual data rows we have collected
        int dataRowCount = 0;

        // --- Start reading the file ---
        // BufferedReader reads the file line by line efficiently
        // FileReader opens the actual file from the path the user gave
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {

            String line; // 'line' holds the current line being read

            // Keep reading lines until there are no more (null means end of file)
            while ((line = br.readLine()) != null) {

                // Skip completely blank lines
                if (line.trim().isEmpty()) continue;

                // Split the line into columns using comma as separator
                // -1 ensures trailing empty values are kept
                String[] columns = line.split(",", -1);

                // --- Find the real header row ---
                // We look for the row that starts with "Candidate"
                if (headerColumns == null && columns[0].trim().equalsIgnoreCase("Candidate")) {
                    headerColumns = columns; // Save the header columns
                    continue;               // Skip to the next line
                }

                // --- Collect data rows that come AFTER the header ---
                // Skip rows where the first column is empty (metadata/footer)
                if (headerColumns != null && !columns[0].trim().isEmpty()) {
                    first10Rows.add(columns); // Add this row to our list
                    dataRowCount++;

                    // Stop reading once we have collected 10 rows
                    if (dataRowCount == 10) break;
                }
            }

            // --- Display the results ---
            System.out.println("\n============================================================");
            System.out.println("           MP02 - FIRST 10 ROWS OF DATASET");
            System.out.println("============================================================");
            System.out.println("File: " + filePath);
            System.out.println("------------------------------------------------------------");

            // If no header was found, the file format may be unexpected
            if (headerColumns == null) {
                System.out.println("  ERROR: Could not find the dataset header row.");
            } else {
                // Print each of the first 10 rows in a labeled format
                for (int i = 0; i < first10Rows.size(); i++) {
                    String[] row = first10Rows.get(i);
                    System.out.println("\n  Record #" + (i + 1));
                    System.out.println("  ---------------------------");

                    // Print each field with its column name as a label
                    for (int j = 0; j < headerColumns.length; j++) {
                        String colName = headerColumns[j].trim();

                        // Only display columns that have a real header name
                        if (!colName.isEmpty()) {
                            // Get the value safely; show "(empty)" if missing
                            String value = (j < row.length) ? row[j].trim() : "(empty)";
                            if (value.isEmpty()) value = "(empty)";
                            System.out.printf("  %-12s : %s%n", colName, value);
                        }
                    }
                }

                System.out.println("\n------------------------------------------------------------");
                System.out.println("  Rows Displayed: " + first10Rows.size());
            }

            System.out.println("============================================================");

        } catch (FileNotFoundException e) {
            // This error happens if the file path is wrong or file does not exist
            System.out.println("ERROR: File not found. Please check the file path.");
        } catch (IOException e) {
            // This catches any other file reading errors
            System.out.println("ERROR: Could not read the file. " + e.getMessage());
        }

        scanner.close(); // Always close the scanner when done
    }
}

File: MIDTERM-LAB-3/MP02/MP02.js
// ============================================================
// MP02 - Display First 10 Rows of Dataset
// Programmer: EYAS, SOPHIA WELYNNE S.
// Course: Programming 2 | BSIT GD - 1st Year
// Description: This program reads a CSV file and displays
//              the first 10 data rows in a readable format,
//              along with the column headers of the dataset.
// ============================================================

// Import the built-in Node.js modules needed for this program
const fs = require("fs");         // 'fs' is used to read files from disk
const readline = require("readline"); // 'readline' is used to get input from the user

// Create an interface to read input from the keyboard (stdin)
// and output to the screen (stdout)
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Ask the user to enter the path to the CSV file
rl.question("Enter the CSV dataset file path: ", (filePath) => {

    // Close the readline interface after getting the input
    rl.close();

    // --- Read the file ---
    // Check first if the file actually exists at the given path
    if (!fs.existsSync(filePath)) {
        console.log("ERROR: File not found. Please check the file path.");
        return; // Stop the program if the file is not found
    }

    // Read the entire file content as a single string (UTF-8 encoding)
    const fileContent = fs.readFileSync(filePath, "utf8");

    // Split the file content into individual lines
    // The filter removes any completely empty lines from the array
    const lines = fileContent.split("\n").filter(line => line.trim() !== "");

    // 'headerColumns' will store the column names from the header row
    let headerColumns = null;

    // 'first10Rows' will hold up to 10 parsed data rows (each row is an array)
    let first10Rows = [];

    // --- Process each line to find header and collect first 10 data rows ---
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim(); // Remove extra whitespace

        // Split the current line into individual column values
        const columns = line.split(",");

        // --- Find the real header row ---
        // We identify it by checking if the first column is "Candidate"
        if (headerColumns === null && columns[0].trim().toLowerCase() === "candidate") {
            headerColumns = columns.map(col => col.trim()); // Save cleaned header names
            continue; // Move to the next line
        }

        // --- Collect data rows that come AFTER the header ---
        // Skip rows where the first column is empty (metadata or footer rows)
        if (headerColumns !== null && columns[0].trim() !== "") {
            first10Rows.push(columns); // Add this row to our collection

            // Stop once we have collected 10 data rows
            if (first10Rows.length === 10) break;
        }
    }

    // --- Display the results ---
    console.log("\n============================================================");
    console.log("           MP02 - FIRST 10 ROWS OF DATASET");
    console.log("============================================================");
    console.log("File: " + filePath);
    console.log("------------------------------------------------------------");

    // If no header was found, the file format may be unexpected
    if (headerColumns === null) {
        console.log("  ERROR: Could not find the dataset header row.");
    } else {
        // Loop through each of the first 10 rows and display them
        first10Rows.forEach((row, index) => {
            console.log(`\n  Record #${index + 1}`);
            console.log("  ---------------------------");

            // Print each field with its column name as a label
            headerColumns.forEach((colName, j) => {

                // Only display columns that have a real header name
                if (colName !== "") {
                    // Get the value safely; show "(empty)" if the field is missing
                    let value = (j < row.length) ? row[j].trim() : "(empty)";
                    if (value === "") value = "(empty)";

                    // Pad the column name for neat alignment
                    console.log(`  ${colName.padEnd(12)} : ${value}`);
                }
            });
        });

        console.log("\n------------------------------------------------------------");
        console.log("  Rows Displayed: " + first10Rows.length);
    }

    console.log("============================================================");
});